website/docs/02-separation-logic-and-biabduction.md

```diff
@@ -0,0 +1,264 @@
+
+## Separation logic
+
+Separation logic is a novel kind of mathematical logic which facilitates reasoning about
+mutations to computer memory. It enables scalability by breaking reasoning into chunks
+corresponding to local operations on memory, and then composing
+the reasoning chunks together.
+
+Separation logic is based on a logical connective <Math code="*" /> called the _separating conjunction_ and pronounced "and separately". Separation logic formulae are interpreted over program allocated heaps. The logical formula
+<Math code="A*B" /> holds of a piece of program heap (a heaplet) when it can be divided into two sub-heaplets described by <Math code="A" /> and <Math code="B" />.
+For example, the formula
+
+---
+
+<Math code="x%20%5Cmapsto%20y%20*%20y%20%5Cmapsto%20x" />
+
+---
+
+can be read "<Math code="x" /> points to <Math code="y" /> and separately <Math code="y" /> points to <Math code="x" />". This formula describes precisely two allocated memory cells. The first cell is allocated at the address denoted by the pointer <Math code="x" /> and the content of this cell is the value of <Math code="y" />.
+The second cell is allocated at the address denoted by the pointer <Math code="y" /> and the content of this second cell is the value of <Math code="x" />. Crucially, we know that there are precisely two cells because <Math code="*" /> stipulates that they are separated and therefore the cells are allocated in two different parts of memory. In other words, <Math code="*" />
+says that <Math code="x" /> and <Math code="y" /> do not hold the same value (i.e., these pointers are not aliased).
+The heaplet partitioning defined by the formula above can visualized like so:
+
+![](/img/SepSplit.jpg)
+
+The important thing about separating conjunction is
+the way that it fits together with mutation to computer memory; reasoning about program commands
+tends to work by updating <Math code="*" />-conjuncts in-place, mimicking the operational in-place update of RAM.
+
+Separation logic uses Hoare triples of the form <Math code="%5Clbrace%20pre%20%5Crbrace%20prog%20%5Clbrace%20post%20%5Crbrace" /> where <Math code="pre" /> is the precondition, <Math code="prog" /> a program part, and <Math code="post" />
+the postcondition. Triples are abstract specifications of the behavior of the program. For example, we could take
+
+---
+
+<Math code="%5Clbrace%20r%20%5Cmapsto%20open%5Crbrace%20%5C%2C%20closeResource(r)%5C%2C%20%5Clbrace%20r%20%5Cmapsto%20closed%5Crbrace%20%20%5C%3B%5C%3B%5C%3B%20(spec)" />
+
+---
+
+as a specification for a method which closes a resource given to it as a parameter.
+
+Now, suppose we have two resources <Math code="r%5C_1" /> and <Math code="r%5C_2" />, described by <Math code="r%5C_1%20%5Cmapsto%20open%20*%20r%5C_2%20%5Cmapsto%20open" />
+and we close the first of them. We think operationally in terms of updating the memory in place, leaving \\(r_2 \mapsto open\\) alone,
```

Comment:
some LaTeX left here that doesn't render anymore

Reply:
Fixed